predicate parsing with OR broken

### Problem description
See https://github.com/JDASoftwareGroup/kartothek/pull/294#issuecomment-638922148

However, after fixing the case for empty predicates on that development branch, the following bug shows up:
Reading datasets using somewhat complex predicates sometimes returns incorrect data.


### Example code (ideally copy-pastable)
```python

from tempfile import TemporaryDirectory
from storefact import get_store_from_url
from functools import partial

store_factory = partial(get_store_from_url, f"hfs://{TemporaryDirectory().name}")
dataset_uuid = "test"

import pandas as pd

df = pd.DataFrame({"A": range(10), "B": ["A", "B"] * 5, "C": range(-10, 0)})

from kartothek.io.eager import read_table, store_dataframes_as_dataset

dm = store_dataframes_as_dataset(
    store=store_factory, dataset_uuid=dataset_uuid, dfs=[df], partition_on=["A", "B"]
)

df1 = read_table(
    store=store_factory,
    dataset_uuid=dataset_uuid,
    predicates=[[("A", "<", 3)], [("A", ">", 5)], [("B", "==", "non-existent")]],
)
print(df1)

df2 = read_table(
    store=store_factory,
    dataset_uuid=dataset_uuid,
    predicates=[[("A", "<", 3)], [("A", ">", 5)]],
)

pd.testing.assert_frame_equal(df1, df2)
```
Output:
```
   A  B   C
0  0  A -10
1  1  B  -9
2  2  A  -8
3  3  B  -7
4  4  A  -6
5  5  B  -5
6  6  A  -4
7  7  B  -3
8  8  A  -2
9  9  B  -1
Traceback (most recent call last):
  File "/Users/xrzq/miniconda3/envs/sc-lastcalcs/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3331, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-d6e03a9f4370>", line 31, in <module>
    pd.testing.assert_frame_equal(df1, df2)
  File "/Users/xrzq/miniconda3/envs/sc-lastcalcs/lib/python3.6/site-packages/pandas/_testing.py", line 1325, in assert_frame_equal
    obj, f"{obj} shape mismatch", f"{repr(left.shape)}", f"{repr(right.shape)}",
  File "/Users/xrzq/miniconda3/envs/sc-lastcalcs/lib/python3.6/site-packages/pandas/_testing.py", line 915, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame are different
DataFrame shape mismatch
[left]:  (10, 3)
[right]: (7, 3)
```
Note that `df1` contains values for `A` outside of the range defined by the predicates.

--- a/kartothek/core/dataset.py
+++ b/kartothek/core/dataset.py
@@ -361,7 +361,7 @@ class DatasetMetadataBase(CopyMixin):
         predicates: PredicatesType = None,
     ):
         """
-        Converts the dataset indices to a pandas dataframe.
+        Converts the dataset indices to a pandas dataframe and filter relevant indices by `predicates`.
 
         For a dataset with indices on columns `column_a` and `column_b` and three partitions,
         the dataset output may look like
@@ -381,22 +381,76 @@ class DatasetMetadataBase(CopyMixin):
         elif columns == []:
             return pd.DataFrame(index=self.partitions)
 
-        dfs = []
         columns_to_scan = columns[:]
         if predicates:
             predicate_columns = columns_in_predicates(predicates)
-            # Don't use set logic to preserve order
-            for col in predicate_columns:
-                if col not in columns_to_scan and col in self.indices:
-                    columns_to_scan.append(col)
+            columns_to_scan = sorted(
+                (predicate_columns & self.indices.keys()) | set(columns)
+            )
 
-        for col in columns_to_scan:
-            if col not in self.indices:
-                if col in self.partition_keys:
-                    raise RuntimeError(
-                        "Partition indices not loaded. Please call `DatasetMetadata.load_partition_keys` first."
-                    )
-                raise ValueError("Index `{}` unknown.")
+            dfs = (
+                self._evaluate_conjunction(
+                    columns=columns_to_scan,
+                    predicates=[conjunction],
+                    date_as_object=date_as_object,
+                )
+                for conjunction in predicates
+            )
+
+            df = pd.concat(dfs)
+            index_name = df.index.name
+            df = (
+                df.loc[:, columns].reset_index().drop_duplicates().set_index(index_name)
+            )
+        else:
+            df = self._evaluate_conjunction(
+                columns=columns_to_scan, predicates=None, date_as_object=date_as_object,
+            )
+        return df
+
+    def _evaluate_conjunction(self, columns, predicates, date_as_object):
+        """
+        Evaluate all predicates related to `columns` to "AND".
+
+        Parameters
+        ----------
+        columns:
+            A list of all columns, including query and index columns.
+        predicates: list of list of tuple[str, str, Any]
+            Optional list of predicates, like [[('x', '>', 0), ...], that are used
+            to filter the resulting DataFrame, possibly using predicate pushdown,
+            if supported by the file format.
+            This parameter is not compatible with filter_query.
+
+            Predicates are expressed in disjunctive normal form (DNF). This means
+            that the innermost tuple describes a single column predicate. These
+            inner predicates are all combined with a conjunction (AND) into a
+            larger predicate. The most outer list then combines all predicates
+            with a disjunction (OR). By this, we should be able to express all
+            kinds of predicates that are possible using boolean logic.
+
+            Available operators are: `==`, `!=`, `<=`, `>=`, `<`, `>` and `in`.
+        dates_as_object: bool
+            Load pyarrow.date{32,64} columns as ``object`` columns in Pandas
+            instead of using ``np.datetime64`` to preserve their type. While
+            this improves type-safety, this comes at a performance cost.
+
+        Returns
+        -------
+        df_result:
+            A DataFrame containing all indices for which `predicates` holds true.
+        """
+        non_index_columns = set(columns) - self.indices.keys()
+        if non_index_columns:
+            if non_index_columns & set(self.partition_keys):
+                raise RuntimeError(
+                    "Partition indices not loaded. Please call `DatasetMetadata.load_partition_indices` first."
+                )
+            raise ValueError(
+                "Unknown index columns: {}".format(", ".join(sorted(non_index_columns)))
+            )
+        dfs = []
+        for col in columns:
             df = pd.DataFrame(
                 self.indices[col].as_flat_series(
                     partitions_as_index=True,
@@ -405,24 +459,16 @@ class DatasetMetadataBase(CopyMixin):
                 )
             )
             dfs.append(df)
-
-        # start joining with the small ones
-        sorted_dfs = sorted(dfs, key=lambda df: len(df))
-        result = sorted_dfs.pop(0)
-        for df in sorted_dfs:
-            result = result.merge(df, left_index=True, right_index=True, copy=False)
-
-        if predicates:
-            index_name = result.index.name
-            result = (
-                result.loc[:, columns]
-                .reset_index()
-                .drop_duplicates()
-                .set_index(index_name)
+        # dfs contains one df per index column. Each df stores indices filtered by `predicates` for each column.
+        # Performing an inner join on these dfs yields the resulting "AND" evaluation for all of these predicates.
+        # We start joining with the smallest dataframe, therefore the sorting.
+        dfs_sorted = sorted(dfs, key=len)
+        df_result = dfs_sorted.pop(0)
+        for df in dfs_sorted:
+            df_result = df_result.merge(
+                df, left_index=True, right_index=True, copy=False
             )
-            return result
-        else:
-            return result
+        return df_result
 
 
 class DatasetMetadata(DatasetMetadataBase):

--- a/kartothek/core/index.py
+++ b/kartothek/core/index.py
@@ -451,13 +451,34 @@ class IndexBase(CopyMixin):
         Parameters
         ----------
         compact:
-            If True, the index will be unique and the Series.values will be a list of partitions/values
+            If True, ensures that the index will be unique. If there a multiple partition values per index, there values
+            will be compacted into a list (see Examples section).
         partitions_as_index:
-            If True, the relation between index values and partitions will be reverted for the output
+            If True, the relation between index values and partitions will be reverted for the output dataframe:
+            partition values will be used as index and the indices will be mapped to the partitions.
         predicates:
             A list of predicates. If a literal within the provided predicates
             references a column which is not part of this index, this literal is
             interpreted as True.
+
+        Examples:
+
+        .. code::
+        >>> index1 = ExplicitSecondaryIndex(
+        ...     column="col", index_dct={1: ["part_1", "part_2"]}, dtype=pa.int64()
+        ... )
+        >>> index1
+            col
+            1    part_1
+            1    part_2
+        >>> index1.as_flat_series(compact=True)
+            col
+            1    [part_1, part_2]
+        >>> index1.as_flat_series(partitions_as_index=True)
+            partition
+            part_1    1
+            part_2    1
+
         """
         check_predicates(predicates)
         table = _index_dct_to_table(

--- a/kartothek/io_components/metapartition.py
+++ b/kartothek/io_components/metapartition.py
@@ -601,6 +601,8 @@ class MetaPartition(Iterable):
         index_df = pd.DataFrame(index_df_dct)
 
         filtered_predicates = []
+        # We assume that indices on the partition level have been filtered out already in `dispatch_metapartitions`.
+        # `filtered_predicates` should only contain predicates that can be evaluated on parquet level
         for conjunction in split_predicates:
             predicates = [conjunction.key_part]
             if (

--- a/tests/io/eager/test_read.py
+++ b/tests/io/eager/test_read.py
@@ -9,6 +9,7 @@ from kartothek.io.eager import (
     read_dataset_as_dataframes,
     read_dataset_as_metapartitions,
     read_table,
+    store_dataframes_as_dataset,
 )
 from kartothek.io.testing.read import *  # noqa
 from kartothek.io_components.metapartition import SINGLE_TABLE
@@ -158,7 +159,6 @@ def test_read_table_simple_list_for_cols_cats(dataset, store_session):
 
 
 def test_write_and_read_default_table_name(store_session):
-    from kartothek.io.eager import store_dataframes_as_dataset
 
     df_write = pd.DataFrame({"P": [3.14, 2.71]})
     store_dataframes_as_dataset(store_session, "test_default_table_name", [df_write])
@@ -168,3 +168,39 @@ def test_write_and_read_default_table_name(store_session):
         "test_default_table_name", store_session
     )
     pd.testing.assert_frame_equal(df_write, df_read_as_dfs[0]["table"])
+
+
+@pytest.mark.parametrize("partition_on", [None, ["A", "B"]])
+def test_read_or_predicates(store_factory, partition_on):
+    # https://github.com/JDASoftwareGroup/kartothek/issues/295
+    dataset_uuid = "test"
+    df = pd.DataFrame({"A": range(10), "B": ["A", "B"] * 5, "C": range(-10, 0)})
+
+    store_dataframes_as_dataset(
+        store=store_factory,
+        dataset_uuid=dataset_uuid,
+        dfs=[df],
+        partition_on=partition_on,
+    )
+
+    df1 = read_table(
+        store=store_factory,
+        dataset_uuid=dataset_uuid,
+        predicates=[[("A", "<", 3)], [("A", ">", 5)], [("B", "==", "non-existent")]],
+    )
+
+    df2 = read_table(
+        store=store_factory,
+        dataset_uuid=dataset_uuid,
+        predicates=[[("A", "<", 3)], [("A", ">", 5)]],
+    )
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 1, 2, 6, 7, 8, 9],
+            "B": ["A", "B", "A", "A", "B", "A", "B"],
+            "C": [-10, -9, -8, -4, -3, -2, -1],
+        },
+    )
+
+    pd.testing.assert_frame_equal(df1, df2)
+    pd.testing.assert_frame_equal(expected, df2)

--- a/tests/io_components/test_read.py
+++ b/tests/io_components/test_read.py
@@ -200,3 +200,82 @@ def test_dispatch_metapartitions_sorted_dispatch_by(store):
             current = current[0][2]
             assert current > last
             last = current
+
+
+def test_dispatch_metapartitions_complex_or_predicates(store_factory):
+    dataset_uuid = "test"
+    df = pd.DataFrame({"A": range(10), "B": ["A", "B"] * 5, "C": range(-10, 0)})
+
+    store_dataframes_as_dataset(
+        store=store_factory,
+        dataset_uuid=dataset_uuid,
+        dfs=[df],
+        partition_on=["A", "B"],
+    )
+    predicates = [[("A", "<", 3)], [("B", "==", "B")]]
+    mps = [
+        mp.load_dataframes(store_factory)
+        for mp in dispatch_metapartitions(
+            dataset_uuid, store_factory, predicates=predicates
+        )
+    ]
+    actual = pd.concat([mp.data["table"] for mp in mps])
+    actual = actual.sort_values(by="A", ignore_index=True)
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 1, 2, 3, 5, 7, 9],
+            "B": ["A", "B", "A", "B", "B", "B", "B"],
+            "C": [-10, -9, -8, -7, -5, -3, -1],
+        }
+    )
+    pd.testing.assert_frame_equal(actual, expected)
+
+    predicates = [[("A", "<", 3)], [("B", "==", "notthere")]]
+    mps = [
+        mp.load_dataframes(store_factory)
+        for mp in dispatch_metapartitions(
+            dataset_uuid, store_factory, predicates=predicates
+        )
+    ]
+    actual = pd.concat([mp.data["table"] for mp in mps])
+    actual = actual.sort_values(by="A", ignore_index=True)
+    expected = pd.DataFrame(
+        data={"A": [0, 1, 2], "B": ["A", "B", "A"], "C": [-10, -9, -8]}
+    )
+    pd.testing.assert_frame_equal(actual, expected)
+
+    predicates = [[("A", "<", 3), ("B", "==", "A")], [("B", "==", "B"), ("A", ">", 2)]]
+    mps = [
+        mp.load_dataframes(store_factory)
+        for mp in dispatch_metapartitions(
+            dataset_uuid, store_factory, predicates=predicates
+        )
+    ]
+    actual = pd.concat([mp.data["table"] for mp in mps])
+    actual = actual.sort_values(by="A", ignore_index=True)
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 2, 3, 5, 7, 9],
+            "B": ["A", "A", "B", "B", "B", "B"],
+            "C": [-10, -8, -7, -5, -3, -1],
+        }
+    )
+    pd.testing.assert_frame_equal(actual, expected)
+
+    predicates = [[("A", "<", 3)], [("B", "==", "B"), ("A", ">", 2)]]
+    mps = [
+        mp.load_dataframes(store_factory)
+        for mp in dispatch_metapartitions(
+            dataset_uuid, store_factory, predicates=predicates
+        )
+    ]
+    actual = pd.concat([mp.data["table"] for mp in mps])
+    actual = actual.sort_values(by="A", ignore_index=True)
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 1, 2, 3, 5, 7, 9],
+            "B": ["A", "B", "A", "B", "B", "B", "B"],
+            "C": [-10, -9, -8, -7, -5, -3, -1],
+        }
+    )
+    pd.testing.assert_frame_equal(actual, expected)

--- a/tests/serialization/test_filter.py
+++ b/tests/serialization/test_filter.py
@@ -209,3 +209,32 @@ def test_filter_df_from_predicates_empty_in(value):
     actual = filter_df_from_predicates(df, predicates)
     expected = df.iloc[[]]
     pdt.assert_frame_equal(actual, expected, check_categorical=False)
+
+
+def test_filter_df_from_predicates_or_predicates():
+    df = pd.DataFrame({"A": range(10), "B": ["A", "B"] * 5, "C": range(-10, 0)})
+
+    predicates = [[("A", "<", 3)], [("A", ">", 5)], [("B", "==", "non-existent")]]
+    actual = filter_df_from_predicates(df, predicates)
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 1, 2, 6, 7, 8, 9],
+            "B": ["A", "B", "A", "A", "B", "A", "B"],
+            "C": [-10, -9, -8, -4, -3, -2, -1],
+        },
+        index=[0, 1, 2, 6, 7, 8, 9],
+    )
+    pdt.assert_frame_equal(actual, expected)
+
+    predicates = [[("A", "<", 3)], [("A", ">", 5)], [("B", "==", "B")]]
+    actual = filter_df_from_predicates(df, predicates)
+    # row for (A == 4) is filtered out
+    expected = pd.DataFrame(
+        data={
+            "A": [0, 1, 2, 3, 5, 6, 7, 8, 9],
+            "B": ["A", "B", "A", "B", "B", "A", "B", "A", "B"],
+            "C": [-10, -9, -8, -7, -5, -4, -3, -2, -1],
+        },
+        index=[0, 1, 2, 3, 5, 6, 7, 8, 9],
+    )
+    pdt.assert_frame_equal(actual, expected)

